### PR TITLE
Eliminate server crash with FP64 input

### LINKED ISSUE #108: Unable to use FP64 data type in input to xgb model
I am encountering the following error when I try to use FP64 data type for input to an xgboost regression model on fil backend. (Using perf_analyzer to send requests)

```I0714 07:15:45.803227 310 infer_request.cc:497] prepared: [0x0x7f0e9800ec90] request id: , model: dummy_xgb_regressor, requested version: -1, actual version: 1, flags: 0x0, correlation id: 0, batch size: 1, priority: 0, timeout (us): 0
original inputs:
[0x0x7f0e9801bd38] input: input__0, type: FP64, original shape: [1,32], batch + shape: [1,32], shape: [32]
override inputs:
inputs:
[0x0x7f0e9801bd38] input: input__0, type: FP64, original shape: [1,32], batch + shape: [1,32], shape: [32]
original requested outputs:
output__0
requested outputs:
output__0

[ERROR] evhtp.c:3101     ODDITY, resuming when not paused?!? :: (errno: None)
double free or corruption (out)
Aborted (core dumped)
```

Input/output config - 
```
input [
 {
    name: "input__0"
    data_type: TYPE_FP64
    dims: [ 32 ]
  }
]
output [
 {
    name: "output__0"
    data_type: TYPE_FP32
    dims: [ 1 ]
  }
]
```

Keeping both input & output FP32, or input FP32 & output FP64 is working fine. 

Also I am able to use numpy.float64 input data on the same model using fil in rapidsai container.  

The model is an example xgboost regression model generated from [here](https://github.com/triton-inference-server/fil_backend/blob/main/Example_Models.md).

### PR DESCRIPTION
**Pull request by wphicks**
Respond with error to non-FP32 input and update README to note that FP32 is the only supported I/O type for the FIL backend at this time.

While attempting to fix the underlying issue here, a double free was revealed that occurred if model execution errored out at a particular point in batch processing. Because the existing implementation was somewhat complicated, it was more difficult to reason about the cause of this double free and ultimately eliminate it. In order to facilitate debugging, the `TRITONBACKEND_ModelInstanceExecute` method was restructured to be more readable and less error-prone. As a result, these changes are more far-reaching than strictly necessary to eliminate the crash but provide improved maintainability and stability.

Close #108. Partially address #34.

### CODE REVIEW

**Comment by wphicks** [comment]
@lowener Don't feel like you have to review this entirely on your own, but I'd love to get your thoughts on this one.

File: src/api.cu [hunk 1]
```diff
@@ -167,128 +170,163 @@ TRITONBACKEND_ModelInstanceExecute(
...
   try {
-    auto instance_state = get_instance_state<ModelInstanceState>(*instance);
-    ModelState* model_state = instance_state->StateForModel();
-    auto target_memory = get_native_memory_for_instance(instance_state->Kind());
+    input_batches = get_input_batches<float>(
+        static_cast<uint32_t>(0), requests, target_memory,
+        instance_state->get_raft_handle());
+  }
+  catch (TritonException& setup_err) {
+    return setup_err.error();
```
Comment (lowener) [inline src/api.cu:189]: In which case do we have to use `release_requests`? I see it used after, but I don't see why it would not result in a double free as well ?

File: src/api.cu [hunk 2]
```diff
@@ -167,128 +170,163 @@ TRITONBACKEND_ModelInstanceExecute(
...
-        raw_requests, raw_requests + request_count);
+  // BEGIN: Lambda definitions for processing individual batches
+
+  /* Given an input batch and output batch, perform prediction on input,
+   * storing results in output and returning timing information. Return value
+   * is a std::optional containing a pair of uint64_t representing nanoseconds
+   * since the epoch to the beginning and end of prediction respectively. If
+   * std::optional is std::nullopt, prediction has failed and timing values
+   * would have no meaning */
+  auto instance_predict = [&](decltype(*input_batches.begin())& input_batch,
```
[the remark this answers is not in the document]
Comment (wphicks) [reply src/api.cu:200]: The generic answer is in order to follow [Almost Always Auto](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/) style, but the more specific advantage that it gives us here is that at this point in the code, we do not actually care what the type of `input_batch` is, only that it is what is stored in `input_batches` and accepted by `instance_state->predict`. If the interface changes later on, and `input_batches` now contained something that could be converted to an `InputBatch<float>` implicitly but which was not an `InputBatch<float>`, we might accidentally start incurring the cost of an unnecessary conversion. Using `decltype` here signals that our only goal is to "hook up the pipes" without insisting on a particular type.

That does raise the question of course of why I explicitly named the type for `output_batch`, and the answer is I have no idea! I'll fix that up.

[the remark this answers is not in the document]
Comment (wphicks) [reply src/api.cu:200]: Oh, I remember why. We don't have an `output_batch` until line 244, and I felt like declaring the lambda earlier made for a cleaner/clearer structure. What do you think? Okay as is, or should I reconsider?

File: src/api.cu (hunk 1 above)
Comment (wphicks) [reply src/api.cu:189]: Since we are returning a non-`nullptr` error, the server releases requests itself, as shown [here](https://github.com/triton-inference-server/server/blob/main/src/backends/backend/triton_model.cc#L219). That being said, this seems to contradict the docs [here](https://github.com/triton-inference-server/backend#tritonbackend_request), so I'll open a docs issue against Triton.

Comment (wphicks) [reply src/api.cu:189]: Opened docs issue here: https://github.com/triton-inference-server/server/issues/3176

File: src/api.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (divyegala) [reply src/api.cu:200]: Otherwise, in lambas, I believe you can use `auto` directly

File: src/triton_fil/model_state.h [hunk 3]
```diff
@@ -53,9 +54,10 @@ class ModelState : public BackendModel {
       const uint64_t version);
 
   size_t num_class();
+  std::vector<int64_t> get_output_shape(std::vector<int64_t> input_shape);
 
  private:
-  size_t num_class_;  // C++17: Use std::optional
+  size_t num_class_;
```
Comment (divyegala) [inline src/triton_fil/model_state.h:60]: `std::size_t`?

File: src/triton_fil/model_state.h [hunk 4]
```diff
@@ -53,9 +54,10 @@ class ModelState : public BackendModel {
       const uint64_t version);
 
   size_t num_class();
```
Comment (divyegala) [inline src/triton_fil/model_state.h:56]: `std::size_t`?

File: src/triton_fil/model_instance_state.cuh [hunk 5]
```diff
@@ -58,6 +59,29 @@ class ModelInstanceState : public BackendModelInstance {
       TRITONBACKEND_ModelInstance* triton_model_instance, const char* name,
       const int32_t device_id);
 
+  template <class Iter>
+  typename std::enable_if_t<std::is_same<
```
Comment (divyegala) [inline src/triton_fil/model_instance_state.cuh:63]: I don't understand why SFINAE is needed here. Since you are essentially doing `std::is_same` for exactly one type, can't that type just be the function overload?

**Review by divyegala — COMMENTED** [review]
Small question about requiring SFINAE in all functions where the `std::is_same` check is performed, as well as piggybacking on Mickael's comment on `decltype` in lambas. Either way, I *believe* `auto` can be used in lambas directly for the type to be inferred when the lambda is actually compiled

File: src/api.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (lowener) [reply src/api.cu:200]: Declaring the lambda here instead of at the same scope than `output_batch` is definitely cleaner, so I'm okay with this solution

[the remark this answers is not in the document]
Comment (wphicks) [reply src/api.cu]: > Otherwise, in lambas, I believe you can use auto directly

D'oh! You're quite right. I think I noticed that we needed the `decltype` at lines 233-234 and then got hung up on using it in all the other lambdas.

In terms of being type-agnostic, my argument is not that that lambda *is* type agnostic but it is "just" a restructuring of the existing code. In the original version, that logic would have been in a context where the type is already known, so moving it into a lambda for clearer/cleaner organization should not introduce the need for a new type declaration.

Speaking more generally, I'd argue that `auto` is not about type agnosticism but about deferring type determination to some other source of truth. The following is not really type agnostic, but it's a sensible use of auto:
```c++
std::vector<int> v = {0, 1, 2, 3, 4};
auto i = std::begin(v);
std::cout << *(i++) << "\n";
```
Here, we care very much about the type of `i`. It must be incrementable, and the return type of that incrementation must be dereferencable, so we are depending on the existence of certain members. The benefit that `auto` brings is that we are deferring to `std::begin` (and  from there to `std::vector`) as the "source of truth" about what `i` should be. If that type doesn't have the members we want, then it will just fail to compile, and we know that our source of truth gave us a type that behaves differently than we expected rather than attempting to insist that it is something else using another source of truth. In this specific case, the lambda is ultimately deferring to `get_input_batches` to tell us the type of an input batch, since that function is the source of those objects.

Sorry, that was an excessively-detailed answer, but it was as much for me to think through this myself as anything :smile:. In an effort to make sure that this code is at least fully self-consistent about how I'm using `auto`, I've tweaked a couple more things to get to full AAA-style in `TRITONBACKEND_ModelInstanceExecute`. If we still don't like the implementation after that change, we can look at a more significant restructuring.

File: src/triton_fil/model_instance_state.cuh (hunk 5 above)
Comment (wphicks) [reply src/triton_fil/model_instance_state.cuh:63]: The `enable_if_t` is checking if the value inside the iterator is a `TRITONBACKEND_Request*`, but we want to enable this for all containers that contain such objects. If we had the function just accept an iterator of a specific type, wouldn't we have to be specific about which container's iterator we're working with?

To be honest, I didn't think about this too much except that this is the way I've generally implemented templates that don't care about the details of the iterator type but do care about the contents of that iterator in the past. Is there a better way to do this and/or did I overthink it here?

File: src/api.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (divyegala) [reply src/api.cu]: Love the detailed answer! I don't have much more to add, you're absolutely right. `auto` defers the type checking, but we're still in compiler error land if an input is tried to be passed to the lambda whose type does not work with the contents of the lambda

File: src/triton_fil/model_instance_state.cuh (hunk 5 above)
Comment (divyegala) [reply src/triton_fil/model_instance_state.cuh:63]: I misread the intention of the SFINAE here, I think the pointer type threw me off here. It's an iterator container, whose each element is a pointer to a `TRITORBACKEND_Request`. That makes much more sense, thanks!

File: src/api.cu [hunk 6]
```diff
@@ -165,130 +168,163 @@ TRITONBACKEND_ModelInstanceExecute(
...
-      std::chrono::steady_clock::now().time_since_epoch().count();
-  std::size_t total_inference_count = 0;
-  std::vector<TRITONBACKEND_Response*> responses;
+  auto all_start_time =
+      uint64_t(std::chrono::steady_clock::now().time_since_epoch().count());
+
+  auto requests = std::vector<TRITONBACKEND_Request*>(
+      raw_requests, raw_requests + request_count);
+
+  auto instance_state = get_instance_state<ModelInstanceState>(*instance);
```
Comment (divyegala) [inline src/api.cu:177]: Could you rephrase this as `auto *instance_state` so it's clear it's a pointer type?

File: src/api.cu [hunk 7]
```diff
@@ -165,130 +168,163 @@ TRITONBACKEND_ModelInstanceExecute(
...
+
+  /* Given start and end iterators for requests, process them and send
+   * responses. Return a std::optional containing a pair of values representing
+   * the time in nanoseconds since the epoch at which computation began and
+   * ended for this batch.
+   *
+   * This lambda is guaranteed to send a response for each request in the batch
+   */
+  auto respond_to_requests = [&](auto& batch,
+                                 decltype(requests.begin()) requests_begin,
```
Comment (divyegala) [inline src/api.cu:237]: why no `auto` here?

**Review by divyegala — APPROVED** [review]
Pre-approving, PR looks mostly good. Just a suggestion and a question

File: src/api.cu (hunk 7 above)
Comment (wphicks) [reply src/api.cu]: No good reason, as it turns out :smile:. I had convinced myself that the type could not properly be deduced down at line 242, but it totally can... if I had remembered to include a `typename` in there! Fixed it, and now we're looking much more pleasantly `auto`ish.
